src/DraupMembershipERC721.sol

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.17;
+
+import {Ownable} from "openzeppelin-contracts/contracts/access/Ownable.sol";
+import {IERC2981, IERC165} from "openzeppelin-contracts/contracts/interfaces/IERC2981.sol";
+import {DefaultOperatorFilterer} from "operator-filter-registry/src/DefaultOperatorFilterer.sol";
+import {ERC721} from "openzeppelin-contracts/contracts/token/ERC721/ERC721.sol";
+import {IERC20} from "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import {MerkleProof} from "openzeppelin-contracts/contracts/utils/cryptography/MerkleProof.sol";
+import {PaddedString} from "draup-utils/src/PaddedString.sol";
+import {IRenderer} from "./IRenderer.sol";
+
+contract DraupMembershipERC721 is ERC721, Ownable, DefaultOperatorFilterer {
+    uint256 public immutable MAX_SUPPLY;
+    uint256 public immutable ROYALTY = 750;
+    bool public transfersAllowed = false;
+    IRenderer public renderer;
+    string public baseTokenURI;
+    uint256 public nextTokenId;
+    bytes32 public merkleRoot;
+
+    // Mapping to track who used their allowlist spot
+    mapping(address => bool) private _claimed;
+
+    constructor(uint256 maxSupply, string memory baseURI) ERC721("Draup Membership", "DRAUP") {
+        MAX_SUPPLY = maxSupply;
+        baseTokenURI = baseURI;
+    }
+
+    error InvalidProof();
+    error AlreadyClaimed();
+    error MaxSupplyReached();
+    error TransfersNotAllowed();
+
+    function mint(bytes32[] calldata merkleProof) public {
+        bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(msg.sender, 1))));
```

Comment:
Someone with more merkle tree experience should confirm this, but if you'er just trying to check that an address is part of the tree, you could probably do:

```
bytes32 leaf = keccak256(bytes.concat(keccak256(msg.sender)));
```

I don't know what the gas savings would be. Probably pretty small.

Reply:
I would suggest:
 
```
bytes32 leaf = keccak256(abi.encodePacked(msg.sender));
```

I've used this in the past and it requires hashing only once.

Reply:
I was curious why the OZ code code recommended double-usage of `keccak256`. They mention that it is to prevent second preimage attacks, which are [best described here][1]. In the naive case, an attacker could provide data and a proof for an intermediate node instead of a leaf node and trick the system into thinking that the proof passes.

However, here you do not have that problem because:

1. Most importantly, the input value (`msg.sender`) is 20 bytes long, while the inputs to an internal hash combine to be 64 bytes long. So a value that could be `msg.sender` could never equal the value for an internal node.
1. The attack can only control the input value by calling from specific addresses. Finding these addresses would be time intensive and the attacker's benefit is negligible: they would only be able to cause an NFT to get sent to one of those random addresses.

In the general case, there are two ways to solve the problem:

1. Ensure that the data in a leaf node could never collide with the data in an internal node. Using different width types for leaves vs internal solves this, as does prefixing internal nodes and leaf nodes with different data.
2. [Ensure that leaves use a different hashing algorithm than the internal hashing algorithm][2] when computing their node hash. So if a user supplies an internal node as a leaf, it will be double-hashed as part of the proof, which returns a different value than when it is single-hashed in the normal internal case.

OZ chooses the second approach and does double-hashing of leaves, while internal nodes are hashed once with `keccak256`.

A few more links that were useful for me:

- [`MerkleProof.sol` source code](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/MerkleProof.sol)
- [A bug report about this kind of vulnerability](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3091)
- [OZ's JS library for merkle proofs](https://github.com/OpenZeppelin/merkle-tree#standard-merkle-trees), which kicked off my whole investigation about why they hash twice

[1]: https://flawed.net.nz/2018/02/21/attacking-merkle-trees-with-a-second-preimage-attack/
[2]: https://crypto.stackexchange.com/questions/2106/what-is-the-purpose-of-using-different-hash-functions-for-the-leaves-and-interna

Reply:
Actually, it might be possible that an internal tree node could be equal to a `msg.sender` value extended with 44 zeroes, and then you would have data collisions between leaves and internal nodes. Someone might want to double check me on that vulnerability as it contradicts a bug report in one of my reference links.

Anyways, very unlikely and double hashing leaves eliminates the problem.

Reply:
If you decide to go with the single-hashed version that banksian suggested above, a reminder that you won't be able to use OZ's JS library `StandardMerkleTree` to generate the proof since it does a [double hash](https://github.com/OpenZeppelin/merkle-tree/blob/master/src/standard.ts#L10).

Lanyard has some example code that uses the single hash approach.
https://lanyard.org/tree/sample